R/utilities-patterns.R

```diff
@@ -0,0 +1,105 @@
+
+#' Modify fill transparency
+#'
+#' This works much like [alpha()][scales::alpha] in that it modifies the
+#' transparency of fill colours. It differs in that is also attempts to set
+#' the transparency of `<GridPattern>` objects.
+#'
+#' @param fill A fill colour given as a `character` or `integer` vector, or as a
+#'   list of `<GridPattern>` objects.
+#' @param alpha A transparency value between 0 (transparent) and 1 (opaque),
+#'   parallel to `fill`.
+#'
+#' @return A `character` vector of colours, or list of `<GridPattern>` objects.
+#' @export
+#' @keywords internal
+#'
+#' @examples
+#' # Typical colour input
+#' fill_alpha("red", 0.5)
+#'
+#' if (utils::packageVersion("grid") > "4.2") {
+#'   # Pattern input
+#'   fill_alpha(list(grid::linearGradient()), 0.5)
+#' }
+fill_alpha <- function(fill, alpha) {
+  if (!is.list(fill)) {
+    # Happy path for no patterns
+    return(alpha(fill, alpha))
+  }
+  if (any(vapply(fill, is_pattern, logical(1)))) {
+    if (utils::packageVersion("grid") < "4.2") {
+      # Pattern fills were introduced in R 4.1.0, but *vectorised* patterns
+      # were only introduced in R 4.2.0.
+      cli::cli_abort(
+        "Using patterns in {.pkg ggplot2} requires at least R version 4.2.0."
+      )
+    }
```

Comment:
Would it make sense to throw a warning if the active graphics device doesn't support these patterns?

Something like the equivalent of `gridpattern::guess_has_R4.1_features(c("gradients", "patterns"))`.  In the future this should simply be `all(c("LinearGradient", "RadialGradient", "TilingPattern") %in% dev.capabilities()$patterns)` once `{ragg}` and `{svglite}` are updated to pass this info to `dev.capabilities()`.

Reply:
Thanks, yeah that is a good suggestion. I've had a look at `guess_has_R4.1_features()` and kudos to you for being so careful to your users to properly check the device's capabilities! I've no issue just rejecting ragg/svglite until they report their capabilities properly though.

Reply:
Adding this info was stalled as implementation of the new composition operators became more complex than expected. I would advice to special case them since these are our (Posits) packages, they do provide the required functionality, and we advise their use throughout

Reply:
> I would advice to special case them since these are our (Posits) packages, they do provide the required functionality, and we advise their use throughout

@thomasp85 what is the recommended way to "special case" whether the active graphics device is `svglite::svglite()` (supports alpha masks and gradients) instead of `vdiffr:::svglite()` (does not support alpha masks and gradients) since for both graphics devices `names(dev.cur())` is `"devSVG"` and `grDevices::dev.capabilities()$patterns` is `NA`.

Reply:
> it would be really nice to have a function(s) exported from ggplot2 that checks for support for different graphics engine features

@mjskay I've refactored `gridpattern::guess_has_R4.1_features()` so it is now also available as a "standalone" file.
  If you don't want to import `{gridpattern}` you may copy its source under the permissive [Unlicense](https://unlicense.org) into your own R package
  by either using `usethis::use_standalone("trevorld/gridpattern", "standalone-guess_has_R4.1_features.R")`
  or simply copying `standalone-guess_has_R4.1_features.R` into your `R` directory and 
  adding `grDevices` and `utils` to the `Imports` of your `DESCRIPTION` file.

Although not perfect it goes beyond using `grDevices::dev.capabilities()` and can guess R 4.1 graphic feature support in most supported base R graphic devices in R version 4.1 as well as in `{ragg}` in R versions 4.1+ and sometimes (but not always) in `{svglite}`.

Known limitations: will conservatively but incorrectly guess `FALSE` for `svglite::svglite()` if called within `vdiffr::expect_doppelganger()` and for `png(type = "cairo")` in R version 4.1 on Windows.

p.s. In case you weren't aware `gridpattern::patternGrob(pattern = "gradient")` can provide linear gradient support (via a rasterized fallback) on graphics devices that don't support the R 4.1 gradient pattern feature including R versions before R 4.1.